src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/VSCodeAnalyzerLoader.cs

```diff
@@ -34,8 +37,34 @@ public void InitializeDiagnosticsServices(Workspace workspace)
         _diagnosticService.Register((IDiagnosticUpdateSource)_analyzerService);
     }
 
-    public static IAnalyzerAssemblyLoader CreateAnalyzerAssemblyLoader()
+    public static IAnalyzerAssemblyLoader CreateAnalyzerAssemblyLoader(ExtensionAssemblyManager extensionAssemblyManager, ILogger logger)
     {
-        return new DefaultAnalyzerAssemblyLoader();
+        return new VSCodeExtensionAssemblyAnalyzerLoader(extensionAssemblyManager, logger);
+    }
+
+    /// <summary>
+    /// Analyzer loader that will re-use already loaded assemblies from the extension load context.
+    /// </summary>
+    /// <param name="extensionAssemblyManager"></param>
+    private class VSCodeExtensionAssemblyAnalyzerLoader(ExtensionAssemblyManager extensionAssemblyManager, ILogger logger) : IAnalyzerAssemblyLoader
+    {
+        private readonly DefaultAnalyzerAssemblyLoader _defaultLoader = new();
+
+        public void AddDependencyLocation(string fullPath)
+        {
+            _defaultLoader.AddDependencyLocation(fullPath);
+        }
+
+        public Assembly LoadFromPath(string fullPath)
+        {
+            var assembly = extensionAssemblyManager.TryLoadAssemblyInExtensionContext(fullPath);
+            if (assembly is not null)
+            {
+                logger.LogTrace("Loaded analyzer {fullPath} from extension context", fullPath);
+                return assembly;
+            }
+
+            return _defaultLoader.LoadFromPath(fullPath);
```

Comment:
So this won't reuse it, right? Because this is going to give it a file path and then reload that _again_ in the default implementation's ALC?